src/snowflake/connector/errors.py

```diff
@@ -66,10 +73,68 @@ def __unicode__(self):
 
+    def generate_telemetry_exception_data(self: 'Error', msg: Optional[str] = None) -> Dict:
+        """Generate the data to send through telemetry."""
+        telemetry_data = {
+            TelemetryField.KEY_DRIVER_TYPE: CLIENT_NAME,
+            TelemetryField.KEY_DRIVER_VERSION: SNOWFLAKE_CONNECTOR_VERSION
+        }
+        if self.sfqid:
+            telemetry_data[TelemetryField.KEY_SFQID] = self.sfqid
+        if self.sqlstate:
+            telemetry_data[TelemetryField.KEY_SQLSTATE] = self.sqlstate
+        if msg:
+            telemetry_data[TelemetryField.KEY_REASON] = SecretDetector.mask_secrets(msg)
+        if self.errno:
+            telemetry_data[TelemetryField.KEY_ERROR_NUMBER] = str(self.errno)
+        if self.msg:
+            telemetry_data[TelemetryField.KEY_ERROR_MESSAGE] = SecretDetector.mask_secrets(self.msg)
+
+        telemetry_data[TelemetryField.KEY_STACKTRACE] = SecretDetector.mask_secrets(self.__traceback__)
+
+        return telemetry_data
+
+    def send_exception_telemetry(self: 'Error',
+                                 connection: Optional['SnowflakeConnection'],
+                                 telemetry_data: Dict[str, str]):
+        """Send telemetry data by in-band telemetry if it is enabled, otherwise send through out-of-band telemetry."""
+        if connection is not None and connection.telemetry_enabled and not connection._telemetry.is_closed:
+            # Send with in-band telemetry
+            telemetry_data[TelemetryField.KEY_TYPE] = TelemetryField.SQL_EXCEPTION
+            telemetry_data[TelemetryField.KEY_EXCEPTION] = self.__class__.__name__
+            ts = get_time_millis()
+            try:
+                connection._log_telemetry(TelemetryData(telemetry_data, ts))
+            except AttributeError:
+                logger.warning(
+                    "Cursor failed to log to telemetry.",
+                    exc_info=True)
+        else:
+            # Send with out-of-band telemetry
+            telemetry_oob = TelemetryService.get_instance()
+            telemetry_oob.log_general_exception(self.__class__.__name__, telemetry_data)
+
+    def exception_telemetry(self: 'Error',
+                            msg: str,
+                            cursor: 'SnowflakeCursor',
+                            connection: 'SnowflakeConnection'):
+        """Main method to generate and send telemetry data for exceptions."""
+        try:
+            telemetry_data = self.generate_telemetry_exception_data(msg)
+            if cursor is not None:
+                self.send_exception_telemetry(cursor.connection, telemetry_data)
+            elif connection is not None:
+                self.send_exception_telemetry(connection, telemetry_data)
+            else:
+                self.send_exception_telemetry(None, telemetry_data)
+        except Exception:
+            # Do nothing but log if sending telemetry fails
+            logger.warning("Sending exception telemetry failed")
```

Comment:
let's not use the warning level. Failing to send the metric is not critical.

Reply:
Okay I'll change them both to info